vllm/model_executor/parallel_utils/custom_all_reduce.py

```diff
@@ -0,0 +1,223 @@
+from contextlib import contextmanager
+from typing import Optional
+
+import torch
+import torch.distributed as dist
+
+from vllm.logger import init_logger
+from vllm.model_executor.parallel_utils.parallel_state import (
+    get_tensor_model_parallel_world_size, get_tensor_model_parallel_rank)
+
+try:
+    from vllm._C import custom_ar
+    import pynvml
+except ImportError:
+    # For AMD GPUs
+    custom_ar = None
+    pynvml = None
+
+logger = init_logger(__name__)
+
+_CA_HANDLE = None
+_IS_CAPTURING = False
+_SUPPORTED_WORLD_SIZES = [2, 4, 6, 8]
```

Comment:
Just curious: Why should the number of GPUs be even? Which part of the code should we fix if we want to support odd number of GPUs?

Reply:
Well I think my kernels do support old #GPUs. I just never test them. Do other parts of vLLM support old number of GPUs (e.g. tensor parallel linear)?